sdks/java/io/hdfs/src/main/java/org/apache/beam/sdk/io/hdfs/HadoopFileSystem.java

```diff
@@ -17,65 +17,224 @@
+        List<Metadata> metadata = new ArrayList<>();
+        for (FileStatus fileStatus : fileStatuses) {
+          if (fileStatus.isFile()) {
+            metadata.add(Metadata.builder()
+                .setResourceId(new HadoopResourceId(fileStatus.getPath().toUri()))
+                .setIsReadSeekEfficient(true)
+                .setSizeBytes(fileStatus.getLen())
+                .build());
+          }
+        }
+        resultsBuilder.add(MatchResult.create(Status.OK, metadata));
+      } catch (IOException e) {
+        resultsBuilder.add(MatchResult.create(Status.ERROR, e));
+      }
+    }
+    return resultsBuilder.build();
   }
 
   @Override
   protected WritableByteChannel create(HadoopResourceId resourceId, CreateOptions createOptions)
       throws IOException {
-    throw new UnsupportedOperationException();
+    return Channels.newChannel(fileSystem.create(resourceId.toPath()));
   }
 
   @Override
   protected ReadableByteChannel open(HadoopResourceId resourceId) throws IOException {
-    throw new UnsupportedOperationException();
+    FileStatus fileStatus = fileSystem.getFileStatus(resourceId.toPath());
+    return new HadoopSeekableByteChannel(fileStatus, fileSystem.open(resourceId.toPath()));
   }
 
   @Override
   protected void copy(
       List<HadoopResourceId> srcResourceIds,
       List<HadoopResourceId> destResourceIds) throws IOException {
-    throw new UnsupportedOperationException();
+    for (int i = 0; i < srcResourceIds.size(); ++i) {
+      // Unfortunately HDFS FileSystems don't support a native copy operation so we are forced
+      // to use the inefficient implementation found in FileUtil which copies all the bytes through
+      // the local machine.
+      //
+      // HDFS FileSystem does define a concat method but could only find the DFSFileSystem
+      // implementing it. The DFSFileSystem implemented concat by deleting the srcs after which
+      // is not what we want. Also, all the other FileSystem implementations I saw threw
+      // UnsupportedOperationException within concat.
+      FileUtil.copy(
+          fileSystem, srcResourceIds.get(i).toPath(),
+          fileSystem, destResourceIds.get(i).toPath(),
+          false,
+          true,
+          fileSystem.getConf());
+    }
   }
 
   @Override
   protected void rename(
       List<HadoopResourceId> srcResourceIds,
       List<HadoopResourceId> destResourceIds) throws IOException {
-    throw new UnsupportedOperationException();
+    for (int i = 0; i < srcResourceIds.size(); ++i) {
+      fileSystem.rename(
+          srcResourceIds.get(i).toPath(),
+          destResourceIds.get(i).toPath());
+    }
   }
 
   @Override
   protected void delete(Collection<HadoopResourceId> resourceIds) throws IOException {
-    throw new UnsupportedOperationException();
+    for (HadoopResourceId resourceId : resourceIds) {
+      fileSystem.delete(resourceId.toPath(), false);
```

Comment:
So mostly for my own information if @lukecwik wouldn't mind filling me in on the expected behaviour of this API I'd appreciate it.

While going over this PR I was reading the javadoc for the parent class and it says "It is allowed but not recommended to delete directories recursively." and I was a bit confused as to what this meant -- do we allow filesystems to delete directories recursively silently if they so chose OR do we mean that users are allowed to use this API to delete directories recursively (in which case we should  set the second param to true).

Sorry if its a silly question @lukecwik and I don't mean to get in the way just trying to get a better understanding.